### DISCUSSION TITLE
Question about OS detection in the index

### DISCUSSION THREAD

**Discussion by ghost** [alire-project/alire#1437] [General]
Hello! I have been bit in the rear end by libgmp once again (not on Fedora this time), so I decided to take matters in my own hands. I am currently on OpenSUSE Tumbleweed, so I tried to add this line (I took florist as an example since they have the same package name:

```
description = "POSIX.5 Ada interface to operating system services"
name = "florist"

maintainers = ["alejandro@mosteo.com"]
maintainers-logins = ["mosteo"]

[[external]]
kind = "system"
   [external.origin."case(distribution)"]
   "debian" = ["libflorist2018-dev"]
   "ubuntu" = ["libflorist2019-dev"]
   "fedora|opensuse" = ["florist-devel"]
```

However, when I modified the corresponding index file and ran `alr show --external florist`:

```
Kind           Description                 Details                    Available
System package 4 candidate system packages Debian: libflorist2018-dev True     
                                           Ubuntu: libflorist2019-dev          
                                           Fedora: florist-devel               
                                           others: unavailable 
```

I also noticed that if I run `alr show --external glut` which has this index file:
```
description = "OpenGL Utility Toolkit"
name = "glut"

maintainers = ["alejandro@mosteo.com"]
maintainers-logins = ["mosteo"]

[[external]]
kind = "system"
origin = ["freeglut3-dev"]
```
I get:
```
Kind           Description                Details               Available
System package 1 candidate system package Debian: freeglut3-dev True     
                                          Ubuntu: freeglut3-dev          
                                          Msys2: freeglut3-dev           
                                          Arch: freeglut3-dev            
                                          Rhel: freeglut3-dev            
                                          Centos: freeglut3-dev          
                                          Fedora: freeglut3-dev          
                                          others: unavailable  
```

So this brings up a few questions:
1. What operating systems/distros are officially supported? My suspicion is that `alire-platforms.ads` has something to do with this.
2. I would've expected OpenSUSE to be there or at least somewhat merged with Fedora since they mostly use the same package names.
3. Why the RHEL/CentOS/Fedora (does it also detect {Rocky,Alma} Linux?) distinction? 
4. If I have just `"fedora"` in the `external.origin."case(distribution)"` field (which isn't very clear to me how it works, hence no. 1), will it also apply for CentOS and RHEL?

Thank you for your attention.

**Comment by mosteo** [comment]
> 1. What operating systems/distros are officially supported? My suspicion is that alire-platforms.ads has something to do with this.

Yes, `alire-platforms.ads` will tell you which OSes/distros Alire is able to distinguish. As for "official" support, we don't have a list of combos we are committed to support. Rather, the ones that exist in our Continuous Integration test are the ones that we are investing the effort to make sure they work as expected. For that, you can check the workflows in both the `alire` and `alire-index` repos. Also, our docker images for testing are at https://github.com/alire-project/dockerfiles . New additions are welcome (with the caveat that something rather obscure that added a lot of maintenance burden might be out of scope for now).

> 2. I would've expected OpenSUSE to be there or at least somewhat merged with Fedora since they mostly use the same package names.

We have Suse in `Alire.Platforms` and support for Zypper in `Alire.Origins.Deployers.System.Zypper`. If there is no Dockers it's because nobody has gotten around to create it.

> 3. Why the RHEL/CentOS/Fedora (does it also detect {Rocky,Alma} Linux?) distinction?

I guess we erred on the side of caution. In the end this is not too important for the external detection system, but more granularity speeds up things. If it could be guaranteed that those always share package names, they could conceivably be merged.

As for Rocky/Alma, I'm not too familiar with them and we don't have them. It's pretty straighforward to add new distros, though, I can guide you if interested.

It would also be conceivably to have a last resort based on detecting the package manager and trying all known package names for other distros, although this could get quite slow. Better than nothing, maybe? This would be a new feature though, not simply a matter of adding a few new enum values here and there like adding a distro is (mostly).

Detection of distro is in `alire-distributions-current__linux.adb` if you're curious.

> 4. If I have just "fedora" in the external.origin."case(distribution)" field (which isn't very clear to me how it works, hence no. 1), will it also apply for CentOS and RHEL?

No, these `case(distro)` in the index are resolved to the one currently detected, and they don't overlap. 

HTH.

**Comment by mosteo** [comment]
BTW I'm converting this kind of general question to a Discussion.

**Comment by ghost** [comment]
It doesn't seem like it's detecting Suse properly. Specifically, when I changed libgmp-external.toml to this:

```
description = "The GNU Multiple Precision Arithmetic Library"
name = "libgmp"
licenses = "LGPL-3.0-only"
maintainers = ["derodat@adacore.com", "chouteau@adacore.com"]
maintainers-logins = ["pmderodat", "Fabien-Chouteau"]

[[external]]
kind = "system"
    [external.origin."case(distribution)"]
    "debian|ubuntu" = ["libgmp-dev"]
    arch = ["gmp"]
    msys2 = ["mingw-w64-x86_64-gmp"]
    homebrew = ["gmp"]
    macports = ["gmp"]
    "fedora|suse" = ["gmp-devel"]

[[external]]
kind = "version-output"
version-command = [ "pkg-config", "--modversion", "gmp" ]
version-regexp = "([\\d\\.]+)"

# pkg-config requires further steps to make packages findable by
# dependencies that Alire does not yet support
available = false
```

I only get
```
Kind           Description                 Details                     Available
System package 8 candidate system packages Debian: libgmp-dev          True     
                                           Ubuntu: libgmp-dev                   
                                           Msys2: mingw-w64-x86_64-gmp          
                                           Arch: gmp                            
                                           Fedora: gmp-devel                    
                                           others: unavailable                  
Executable     pkg-config --modversion gmp ([\d\.]+)                   False 
```

I am in the process of filling external dependencies for RPM-based distros, maybe I can add `suse` preemptively. 

P.S.: Where's `alire-distributions-current__linux.adb`? I assume you made a typo there
